### ISSUE TITLE
[BUG] Not able to convert date data back to pandas

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/cudf#602]
**Describe the bug**
We don't seem to be able to roundtrip datetime data from pandas to csv to cudf

**Steps/Code to reproduce bug**

```python
import pandas as pd
pdf = pd.util.testing.makeTimeDataFrame().reset_index()[['index', 'A']]
pdf.to_csv('foo.csv')

gdf = cudf.read_csv('foo.csv', names=['index', 'A'], dtype=['date', 'float64'])
gdf.to_pandas()
```

I've tried other dtypes than "date" and read_csv doesn't seem to support them. 

We get the following traceback from the `to_pandas` call

<details>

```python-traceback

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-32-24580ab6f6b0> in <module>
      1 gdf = cudf.read_csv('foo.csv', names=['index', 'A'], dtype=['date', 'float64'])
----> 2 gdf.to_pandas()

~/cudf/python/cudf/dataframe/dataframe.py in to_pandas(self)
   1474         """
   1475         index = self.index.to_pandas()
-> 1476         data = {c: x.to_pandas(index=index) for c, x in self._cols.items()}
   1477         return pd.DataFrame(data, columns=list(self._cols), index=index)
   1478 

~/cudf/python/cudf/dataframe/dataframe.py in <dictcomp>(.0)
   1474         """
   1475         index = self.index.to_pandas()
-> 1476         data = {c: x.to_pandas(index=index) for c, x in self._cols.items()}
   1477         return pd.DataFrame(data, columns=list(self._cols), index=index)
   1478 

~/cudf/python/cudf/dataframe/series.py in to_pandas(self, index)
    496         if index is True:
    497             index = self.index.to_pandas()
--> 498         s = self._column.to_pandas(index=index)
    499         s.name = self.name
    500         return s

~/cudf/python/cudf/dataframe/datetime.py in to_pandas(self, index)
    144 
    145     def to_pandas(self, index):
--> 146         return pd.Series(self.to_array().astype(self.dtype), index=index)
    147 
    148     def to_arrow(self):

~/miniconda/lib/python3.6/site-packages/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    248                                        raise_cast_failure=True)
    249 
--> 250                 data = SingleBlockManager(data, index, fastpath=True)
    251 
    252         generic.NDFrame.__init__(self, data, fastpath=True)

~/miniconda/lib/python3.6/site-packages/pandas/core/internals.py in __init__(self, block, axis, do_integrity_check, fastpath)
   4115         if not isinstance(block, Block):
   4116             block = make_block(block, placement=slice(0, len(axis)), ndim=1,
-> 4117                                fastpath=True)
   4118 
   4119         self.blocks = [block]

~/miniconda/lib/python3.6/site-packages/pandas/core/internals.py in make_block(values, placement, klass, ndim, dtype, fastpath)
   2717                      placement=placement, dtype=dtype)
   2718 
-> 2719     return klass(values, ndim=ndim, fastpath=fastpath, placement=placement)
   2720 
   2721 # TODO: flexible with index=None and/or items=None

~/miniconda/lib/python3.6/site-packages/pandas/core/internals.py in __init__(self, values, placement, fastpath, **kwargs)
   2222 
   2223         super(DatetimeBlock, self).__init__(values, fastpath=True,
-> 2224                                             placement=placement, **kwargs)
   2225 
   2226     def _astype(self, dtype, mgr=None, **kwargs):

~/miniconda/lib/python3.6/site-packages/pandas/core/internals.py in __init__(self, values, placement, ndim, fastpath)
    113             raise ValueError('Wrong number of items passed %d, placement '
    114                              'implies %d' % (len(self.values),
--> 115                                              len(self.mgr_locs)))
    116 
    117     @property

ValueError: Wrong number of items passed 30, placement implies 31
```

</details>

I'm using the latest release, 0.4.0.

**Comment by kkraus14** [comment]
Could you try this on branch-0.5 and report back? The CSV reader has had significant improvements / feature additions / bug fixes since the 0.4 release.

**Comment by mrocklin** [comment]
My attempts to build branch-0.5 locally have failed.  I'll raise a separate issue about that in a bit.

**Comment by mrocklin** [comment]
This continues to fail on branch-0.5

```python
import cudf
cudf.__version__
# '0.4.0+381.g5ade4ab'

import pandas as pd
pdf = pd.util.testing.makeTimeDataFrame().reset_index()[['index', 'A']]
pdf.to_csv('foo.csv')

gdf = cudf.read_csv('foo.csv', names=['index', 'A'], dtype=['date', 'float64'])
gdf.to_pandas()
```

<details>

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-38172bfa6622> in <module>
      4
      5 gdf = cudf.read_csv('foo.csv', names=['index', 'A'], dtype=['date', 'float64'])
----> 6 gdf.to_pandas()

~/cudf/python/cudf/dataframe/dataframe.py in to_pandas(self)
   1567         """
   1568         index = self.index.to_pandas()
-> 1569         data = {c: x.to_pandas(index=index) for c, x in self._cols.items()}
   1570         return pd.DataFrame(data, columns=list(self._cols), index=index)
   1571

~/cudf/python/cudf/dataframe/dataframe.py in <dictcomp>(.0)
   1567         """
   1568         index = self.index.to_pandas()
-> 1569         data = {c: x.to_pandas(index=index) for c, x in self._cols.items()}
   1570         return pd.DataFrame(data, columns=list(self._cols), index=index)
   1571

~/cudf/python/cudf/dataframe/series.py in to_pandas(self, index)
    498         if index is True:
    499             index = self.index.to_pandas()
--> 500         s = self._column.to_pandas(index=index)
    501         s.name = self.name
    502         return s

~/cudf/python/cudf/dataframe/datetime.py in to_pandas(self, index)
    144
    145     def to_pandas(self, index):
--> 146         return pd.Series(self.to_array().astype(self.dtype), index=index)
    147
    148     def to_arrow(self):

~/miniconda/envs/cudf_dev/lib/python3.5/site-packages/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    248                                        raise_cast_failure=True)
    249
--> 250                 data = SingleBlockManager(data, index, fastpath=True)
    251
    252         generic.NDFrame.__init__(self, data, fastpath=True)

~/miniconda/envs/cudf_dev/lib/python3.5/site-packages/pandas/core/internals.py in __init__(self, block, axis, do_integrity_check, fastpath)
   4115         if not isinstance(block, Block):
   4116             block = make_block(block, placement=slice(0, len(axis)), ndim=1,
-> 4117                                fastpath=True)
   4118
   4119         self.blocks = [block]

~/miniconda/envs/cudf_dev/lib/python3.5/site-packages/pandas/core/internals.py in make_block(values, placement, klass, ndim, dtype, fastpath)
   2717                      placement=placement, dtype=dtype)
   2718
-> 2719     return klass(values, ndim=ndim, fastpath=fastpath, placement=placement)
   2720
   2721 # TODO: flexible with index=None and/or items=None

~/miniconda/envs/cudf_dev/lib/python3.5/site-packages/pandas/core/internals.py in __init__(self, values, placement, fastpath, **kwargs)
   2222
   2223         super(DatetimeBlock, self).__init__(values, fastpath=True,
-> 2224                                             placement=placement, **kwargs)
   2225
   2226     def _astype(self, dtype, mgr=None, **kwargs):

~/miniconda/envs/cudf_dev/lib/python3.5/site-packages/pandas/core/internals.py in __init__(self, values, placement, ndim, fastpath)
    113             raise ValueError('Wrong number of items passed %d, placement '
    114                              'implies %d' % (len(self.values),
--> 115                                              len(self.mgr_locs)))
    116
    117     @property

ValueError: Wrong number of items passed 30, placement implies 31
```

</details>

**Comment by kkraus14** [comment]
Looks like you're getting a null in the `index` column which is a datetime64[ms] dtype. Looks like the `to_array` call we're using isn't filling in `NaT` values and is instead dropping the values. Likely just need to add `fillna='pandas'` to that `to_array` call and it should be resolved.

**Comment by vuule** [comment]
The file you generated includes a header row. When you manually specify the column names, the header row is set to None, and this behavior is documented in the csv.py. Because of this, the parser reads the first row as a data row and returns garbage. I see similar behavior with pandas.
You can get the expected behavior with one of the following changes:
1. Specify skiprows=1;
 2. Remove the header row in the file;
3. Not specify the names parameter.

**Comment by kkraus14** [comment]
Fixed by #686.
